### PR TITLE
Test hot/cold splitting in runtime-coreclr crossgen2 pipeline

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Adding testing of hot/cold splitting in Crossgen2 to CI, since Crossgen2 supports this feature on x64.

### CODE REVIEW

**Comment by amanasifkhalid** [comment]
@trylek @ivdiazsa PTAL. Right now, the new hot/cold splitting flag is being passed to `src/tests/build.cmd` in the "build managed test components" step under "coreclr Common Pri0 Test Build AnyOS AnyCPU checked". The managed tests are then downloaded and unzipped by the platform-specific jobs afterwards; I'm guessing the "build managed test components" step, the tests with <AlwaysUsedCrossgen2> set to true will invoke Crossgen with hot/cold splitting enabled. Does this workflow seem right? (Also, I feel like adding a new setting to `src/tests/build.cmd` may not be ideal. Currently, we turn hot/cold splitting on in Crossgen by passing the `--hot-cold-splitting` flag from the command line, so I assumed this was the only way to do it. Please let me know if there's a cleaner way.)

As for the test failures on linux x64, these failures seem to be pre-existing issues [blocking previous runs](https://dev.azure.com/dnceng-public/public/_build?definitionId=124&_a=summary) of the pipeline -- are these failures expected for now?

Also, please ignore the huge diff in `runtime.yml`. I deleted its jobs for now just to speed up CI. I'll make sure to restore everything before opening this draft PR up for reviewing/merging.

Thank you!

**Comment by trylek** [comment]
@amanasifkhalid - Can you please explain to me why we need to pass the hot/cold flag to Roslyn compilation that takes place in managed component build step? I was under the impression that this option is only needed when running Crossgen2 i.o.w. in the ultimate Helix runs. If you're concerned about the execution scripts, you don't need to be, these do get generated at managed component build time but they are recreated upon copying native components to the test tree in the test run phase so that they are updated based on the options for the particular test run.

**Comment by trylek** [comment]
In fact, please keep in mind that Crossgen2 is actually run on the final execution Helix machines so that the only way to instruct its behavior is via the execution scripts and / or environment variables that get passed to Helix (the way it works in practice is that the execution script calls another script emitted by the test infra that contains the "set" / "export" statements setting up the environment variables on the execution machine).

File: eng/pipelines/common/variables.yml
Comment (trylek) [inline eng/pipelines/common/variables.yml:42]: Is this change intentional? It doesn't seem related to what you're trying to achieve.

Comment (amanasifkhalid) [reply eng/pipelines/common/variables.yml:42]: This wasn't related to enabling hot/cold splitting, but just to skip these pipelines for now so they don't keep rerunning whenever I push a commit to this PR

Comment (trylek) [reply eng/pipelines/common/variables.yml:42]: OK, I see, it's just a temporary instrumentation, we just need to make sure to remove it before merging in. I think a common practice is to mark PRs using such forms of temporary instrumentation as "WIP: xxx" (in the title) or as "NO-MERGE" (as a label), once we mark a PR ready for review we and / or the reviewers should make sure there are no remnants of such instrumentations left in the change. We hit this all the time including myself, it's kind of amazing how various logging outputs and other instrumentation changes can sneak into the final change that gets merged in, sometimes leading to surprises. Look at the amount of spew the ```src\tests\build``` command outputs for instance, I believe at least half of this is overlooked or unfixed msbuild / Roslyn spew.

**Comment by amanasifkhalid** [comment]
@trylek Ah sorry, I assumed Crossgen would be invoked in the test build step. So I'm guessing I have to instead pass along some flag enabling hot/cold splitting in the "Send to Helix" step, and then edit the script it uses to handle that flag correctly. I see in `runtime\eng\pipelines\common\templates\runtime\send-to-helix-step.yml`, we pass an "environment" parameter (with switches like "_RunCrossGen2", "_TieringTest", etc.) to `send-to-helix-inner-step.yml`, which then calls `build.ps1` (in the same directory). I'm guessing that environment parameter also needs a hot/cold splitting switch, but I don't see any logic to handle the environment variables passed from `send-to-helix-step.yml` anywhere after (and I also see a warning in the directory's README saying not to edit any of these files, since they can be overwritten by automation) -- do you know where I should be looking? I'm guessing we'll need some logic in the tests' proj files that adds the `--hot-cold-splitting` flag to the Crossgen command if it detects the hot/cold splitting environment variable, but I'm not sure where the logic for that should be.

File: eng/pipelines/coreclr/crossgen2.yml
Comment (ivdiazsa) [inline eng/pipelines/coreclr/crossgen2.yml:41]: We will want later on (in this PR) to add another job. We don't want to get rid of our default Crossgen2 pipelines.

File: eng/pipelines/runtime.yml
Comment (ivdiazsa) [inline eng/pipelines/runtime.yml:63]: Why are we deleting all this?

File: src/tests/build.sh
Comment (ivdiazsa) [inline src/tests/build.sh:103]: We don't need to add Hot/Cold Splitting here. This is just for the normal build and we want it in the test builds, i.e. in `src/tests/build.sh`. As for how to handle the logic of how the test scripts are generated, take a look at `src/tests/Common/helixpublishwitharcade.proj` and `CLRTest.Execute.Bash.targets`. It's all defined there. It's hard to explain from here so feel free to reach out to me offline for more specific guidance on how they work :)

**Review by ivdiazsa — CHANGES_REQUESTED** [review]

File: src/tests/build.sh
Comment (ivdiazsa) [reply src/tests/build.sh:103]: Actually never mind. I thought this was `./build.sh`. Leaving the comment open for visibility on the explanation on the next steps.

File: eng/pipelines/runtime.yml
Comment (amanasifkhalid) [reply eng/pipelines/runtime.yml:63]: This was just a quick-and-dirty way of disabling the runtime pipeline so it doesn't run every time I push to this PR. I'll make sure to restore `runtime.yml` before removing the NO-MERGE label.

Comment (ivdiazsa) [reply eng/pipelines/runtime.yml:63]: Oh okok that's fair.

File: eng/pipelines/coreclr/crossgen2.yml
Comment (amanasifkhalid) [reply eng/pipelines/coreclr/crossgen2.yml:41]: Gotcha, while this PR is a draft, I'll start with enabling hot/cold splitting for the default Crossgen2 pipeline. Once that's working, I'll move this logic to a separate job.

**Comment by trylek** [comment]
@amanasifkhalid - The most important thing to keep in mind is that it's the generated execution script that ultimately runs the test in Helix (possibly including Crossgen2 execution as a preparatory step). Logic around running Crossgen2 resides in the script

https://github.com/dotnet/runtime/blob/main/src/tests/Common/CLRTest.CrossGen.targets

Sadly it has two versions as you can easily imagine, one for Windows and one for Linux. You'll need to audit the Crossgen2 excutions to see whether we already have something we may be able to reuse (some ```Crossgen2ExtraArgs```, some scripts have that notion, I don't remember what exactly the situation is here) and, if not, add it. I think I would slightly prefer to add something generic we can reuse the next time (some Crossgen2ExtraArgs like we use in the .NET SDK instead of a tailored version for just the hot/cold splitting flag) and there are two places I can find as a next step that communicate the environment variables between the Azure job and the Helix execution machine.

File: eng/pipelines/coreclr/crossgen2-hot-cold-splitting.yml
Comment (ivdiazsa) [inline eng/pipelines/coreclr/crossgen2-hot-cold-splitting.yml:1]: Adding a completely new pipeline would probably incur in unwanted additional costs. Just add the hot/cold splitting job to the current Crossgen2 pipeline (i.e. `eng/pipelines/coreclr/crossgen2.yml`

**Review by ivdiazsa — CHANGES_REQUESTED** [review]

**Comment by amanasifkhalid** [comment]
@trylek @ivdiazsa Thank you for bearing with me, and sorry for the azure-pipelines spam! I added a separate job for testing hot/cold splitting, which propagates a flag to send-to-helix-step.yml, which is then used to set an env var in helixpublishwitharcade.proj. Then in CLRTest.CrossGen.targets, the bash/batch scripts add the --hot-cold-splitting flag to Crossgen2ExtraArgs if the env var is set.

File: eng/pipelines/coreclr/crossgen2.yml
Comment (ivdiazsa) [inline eng/pipelines/coreclr/crossgen2.yml:75]: I would change this to `R2R_CG2_HotCold_Splitting` or something like that. Just to make it clear what we are exactly testing there.

Comment (amanasifkhalid) [reply eng/pipelines/coreclr/crossgen2.yml:75]: Gotcha, I'll wait for this pipeline run to finish and then push an update

**Comment by ivdiazsa** [comment]
> @trylek @ivdiazsa Thank you for bearing with me, and sorry for the azure-pipelines spam! I added a separate job for testing hot/cold splitting, which propagates a flag to send-to-helix-step.yml, which is then used to set an env var in helixpublishwitharcade.proj. Then in CLRTest.CrossGen.targets, the bash/batch scripts add the --hot-cold-splitting flag to Crossgen2ExtraArgs if the env var is set.

Don't worry about it, it's all good! The CI isn't exactly the most friendly system to work with so you definitely need to look up help and documentation, and experiment with it. At first glance, this PR seems it's going to work well, but let's wait for the CI to return the results. Also if Tomas has any additional comments.

**Comment by amanasifkhalid** [comment]
Pipeline run isn't finished yet, but I noticed in the "Publish Logs" step of the hot/cold splitting job, the TestRunLogs artifact isn't given a unique name for the new job, so it clashes with the artifact uploaded when running crossgen2 innerloop normally. In the next push, I'll update the name so this doesn't happen.

**Comment by amanasifkhalid** [comment]
@trylek @ivdiazsa Here's a [link](https://dev.azure.com/dnceng-public/public/_build/results?buildId=389509&view=results) to the pipeline results (we can ignore the warnings in the "Publish logs" step for now; my next push will fix that). The failures on linux x64 without splitting seem to be present in previous pipeline runs too, but it looks like the failed tests for linux x64 with splitting are different. Other than that, Windows and OS X are clean with and without splitting. Should we move forward with enabling testing with hot/cold splitting, and just open issues for these failures on Linux (if there aren't issues for them already)? If so, I can push a change addressing Ivan's suggestions and restoring the removed pipelines, and convert this draft into a real PR. Thank you!

**Comment by trylek** [comment]
@amanasifkhalid - We should at least try to make sure that the outerloop pipeline is green with this change, otherwise it will get reverted very quickly. The Crossgen2 pipelines are known not to be completely clean these days, that's also something we should address sooner rather than later.

**Comment by trylek** [comment]
Hmm, never mind, I forgot that you're not enabling hot-cold splitting for the outerloop pipeline.

**Comment by trylek** [comment]
I think that temporarily breaking the crossgen2 pipelines is fine.

**Comment by ivdiazsa** [comment]
@amanasifkhalid I think it's good enough to move the PR from draft to ready for review. Those R2R failures are unrelated to your changes. I also checked the logs to ensure we are enabling hot/cold splitting and indeed it's functioning properly. I just have a little bit more feedback. In your next commit where you reenable all the remaining pipelines and address my remaining comments, I think it would be good if you added the `--hot-cold-splitting` to the tests' response files, rather than passing the flag directly. We've kept it that way for the sake of ease of reading and cleanliness, as well as easier debugging.

**Review by ivdiazsa — APPROVED** [review]
Thanks for doing all this work Aman! Assuming everything passes well in the lab, then it lgtm!

**Comment by amanasifkhalid** [comment]
Sorry please don't approve yet, I pushed some unwanted changes -- fixing them now.

**Comment by ivdiazsa** [comment]
You're making changes to the test scripts themselves, so I triggered an outerloop run just to ensure no corner cases or stuff goes wrong there.

**Comment by amanasifkhalid** [comment]
@ivdiazsa Sorry about pushing in the middle of your pipeline run -- I messed up my local git branch, so when I previously pushed, a couple of old unwanted changes got included. I've fixed that now, so I'll restart your run and re-open for review. Thank you!

**Review by ivdiazsa — APPROVED** [review]

**Comment by amanasifkhalid** [comment]
@trylek do you think this is ready to merge? Thank you!

**Review by trylek — APPROVED** [review]
Looks great to me, thanks Aman!

**Pull request merged**
